Replace OMR_ENV_OPENXL with OMR_TOOLCHAIN=openxl

midronij opened the pull request:
This contribution includes the following changes:

- ~Define `OMR_ENV_OPENXL` for UMA builds~
- ~Remove `OMR_ENV_OPENXL` from CMake builds~
- ~Clean up uses of `OMR_ENV_OPENXL` elsewhere (i.e.: in other config/make files)~
- Replace all usages of `OMR_ENV_OPENXL` with the already existing env variable, `OMR_TOOLCHAIN`

midronij commented:
@keithc-ca here are the changes regarding the usage of `OMR_ENV_OPENXL`. To ensure that it is properly defined for UMA builds, I used another env variable, `OMR_ENV_DATA64`, as a reference, but please let me know if I missed something or if you have any other concerns that I have not yet addressed.

There are also some changes on the OpenJ9 side that I will open a PR for shortly.

@zl-wang FYI

keithc-ca commented on example/glue/configure_includes/configure_aix_ppc.mk:
```diff
@@ -60,6 +60,12 @@ ifeq (aix_ppc, $(SPEC))
     --enable-OMR_THR_FORK_SUPPORT
 endif
 
+ifeq (,$(findstring ibm-clang,$(notdir $(CC))))
```
I think you mean `ifneq` (instead of `ifeq`) here.

keithc-ca commented on example/glue/configure_includes/configure_aix_ppc.mk:
```diff
@@ -74,5 +80,4 @@ CONFIGURE_ARGS += 'AR=ar'
 CONFIGURE_ARGS += 'OMR_HOST_ARCH=ppc'
-CONFIGURE_ARGS += 'OMR_TARGET_DATASIZE=$(TEMP_TARGET_DATASIZE)'
-CONFIGURE_ARGS += 'OMR_TOOLCHAIN=xlc'
+CONFIGURE_ARGS += 'OMR_TARGET_DATASIZE=$(TEMP_TARGET_DATASIZE)'
```
Please don't remove the final newline from text files.

midronij commented:
@keithc-ca as requested, I replaced the all uses of `OMR_ENV_OPENXL` with `OMR_TOOLCHAIN=openxl`, both here and on the matching OpenJ9 PR. Please let me know if there is anything else that needs to be done before this change is ready to be merged

keithc-ca commented on tools/configure:
```diff
@@ -3061,7 +3061,7 @@ return f (e, argv, 0) != argv[0]  ||  f (e, argv, 1) != argv[1];
 }
 _ACEOF
-if [ "$OMR_ENV_OPENXL" == "1" ]; then
+if [[ "$CC" =~ 'ibm-clang' ]]; then
```
This file should be derived from `tools/configure.ac`.

midronij commented:
@keithc-ca I've finished addressing the review comments you've left so far! Please let me know if there are any other changes I should make or concerns I need to address before this can get merged

keithc-ca commented on configure.ac:
```diff
@@ -47,7 +47,7 @@ AC_ARG_VAR([exe_output_dir], [The output directory for executables. (Default: $t
 AC_ARG_VAR([OMR_HOST_ARCH], [The architecture of the CPU where the package will run. One of: aarch64,arm,ppc,riscv,s390,x86])
 AC_ARG_VAR([OMR_TARGET_DATASIZE], [Specifies whether the package will run in 32- or 64-bit mode. One of: 31,32,64])
-AC_ARG_VAR([OMR_TOOLCHAIN], [The toolchain used to build the package. One of: gcc,xlc,msvc])
+AC_ARG_VAR([OMR_TOOLCHAIN], [The toolchain used to build the package. One of: gcc,xlc,openxl,msvc])
```
Can you list the supported toolchains alphabetically, please?

keithc-ca commented on tools/configure:
```diff
@@ -663,7 +663,6 @@ infodir
 docdir
 oldincludedir
 includedir
-runstatedir
```
Where is the (source) change to `tools/configure.ac`?

midronij replied:
After updating `omrmakefiles/confighelpers.m4` to check for OpenXL, I reran autoconf to generate an updated version of `configure` and `tools/configure` (since, as you mentioned, `configure` is supposed to be derived from `configure.ac`, which uses `omrmakefiles/confighelpers.m4`). However, it seems that this particular change occurs even with the previous version of  `omrmakefiles/confighelpers.m4`. Perhaps the `tools/configure` hadn't been updated to reflect the most recent changes to `tools/configure.ac`?

I included those changes in this PR, since as you mentioned, the `configure` files are supposed to be derived from the `configure.ac` files (both of which use  `omrmakefiles/confighelpers.m4`). However, if that's not what we want, then please let me know and I will remove them.

keithc-ca replied:
Please use a newer version of `autoconf` to avoid removing the `runstatedir` portions.

keithc-ca commented on configure.ac:
```diff
@@ -230,6 +230,7 @@ fi
 AS_CASE([$OMR_TOOLCHAIN],
 	[gcc],[],
 	[xlc],[],
+	[openxl],[],
 	[msvc],[],
```
Perhaps these should be sorted to match line 50?

keithc-ca commented on omrmakefiles/confighelpers.m4:
```diff
@@ -67,7 +67,10 @@ AC_DEFUN([OMRCFG_CATEGORIZE_ARCH],
 	[AC_LANG([C])
 	AS_IF([test "x$$1" = "x"],
-		AS_IF([test "$GCC" = "yes"], [$1=gcc]))
+		AS_IF([test "$GCC" = "yes"],
+			AS_IF(AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#ifndef __open_xl__
+#error not an openxl compiler
+#endif]])], [$1=openxl], [$1=gcc]))))
```
Please explain this: Why should we test for OpenXL if `$GCC` is `yes`?

midronij replied:
I'm not sure exactly how the value of `$GCC` is determined, but it seems that with clang-based compilers (like OpenXL or xlclang) it ends up being set to "yes". So, before this change, when using OpenXL, `OMR_TOOLCHAIN` would end up being set to `gcc`. 

I initially put the OpenXL test in there to make sure that the difference between the `openxl` and `gcc` toolchains is caught, but I've now separated it out. This should still successfully detect when OpenXL is being used while being laid out in a more intuitive way.

keithc-ca commented on omrmakefiles/rules.aix.mk:
```diff
@@ -57,28 +57,24 @@ else
+  # openxl options
+  GLOBAL_CFLAGS+=-q mbcs -std=c89 -qinfo=pro
+  GLOBAL_CFLAGS+=-std=c++11
+else ifeq (,$(findstring xlclang,$(notdir $(CC))))
+  # xlc options
+  GLOBAL_CFLAGS+=-q mbcs -qlanglvl=extended -qinfo=pro
 else
   # xlclang options
   GLOBAL_CFLAGS+=-qlanglvl=extended0x -qxlcompatmacros
 endif
```
Suggest inverting this (note the use of `ifneq` instead of `ifeq`):
```
else ifneq (,$(findstring xlclang,$(notdir $(CC))))
  # xlclang options
  GLOBAL_CFLAGS+=-qlanglvl=extended0x -qxlcompatmacros
else
  # xlc options
  GLOBAL_CFLAGS+=-q mbcs -qlanglvl=extended -qinfo=pro
endif
```

keithc-ca commented on omrmakefiles/rules.aix.mk:
```diff
@@ -57,28 +57,24 @@ else
+ifeq (openxl,$(OMR_TOOLCHAIN))
+  # openxl options
+  GLOBAL_CXXFLAGS+=-fno-exceptions
+else ifeq (,$(findstring xlclang++,$(notdir $(CXX))))
+  # xlc++ options
+  GLOBAL_CXXFLAGS+=-q mbcs -qinfo=pro
 else
   # xlclang++ options
   GLOBAL_CXXFLAGS+=-qxlcompatmacros -fno-exceptions
```
Please invert as above.

keithc-ca commented on configure:
```diff
@@ -4,7 +4,6 @@
 #
 #
 # Copyright (C) 1992-1996, 1998-2012 Free Software Foundation, Inc.
-# Copyright IBM Corp. and others 2015
```
We might want to keep this line.
Is the rest of this file automatically generated?

midronij replied:
Yes, this is the `configure` file that autoconf generates with the changes I've made to `configure.ac` and `omrmakefiles/confighelpers.m4`

keithc-ca commented on configure:
```diff
@@ -4685,6 +4698,8 @@ case $OMR_TOOLCHAIN in #(
      ;; #(
   xlc) :
      ;; #(
+  openxl) :
```
I think this file is out-of-date relative to `configure.ac`.

keithc-ca commented on omrmakefiles/rules.aix.mk:
```diff
@@ -28,7 +28,7 @@ ranlib $@
 
 ifeq ($(OMR_ENV_DATA64),1)
-  ifeq ($(OMR_ENV_OPENXL),1)
+  ifeq (openxl,$(OMR_TOOLCHAIN))
```
This is missing 32-bit support with OpenXL.

keithc-ca commented on omrmakefiles/rules.aix.mk:
```diff
@@ -57,28 +57,24 @@ else
+  # openxl options
+  GLOBAL_CFLAGS+=-q mbcs -std=c89 -qinfo=pro
+  GLOBAL_CFLAGS+=-std=c++11
+else ifneq (,$(findstring xlclang,$(notdir $(CC))))
+  # xlc options
+  GLOBAL_CFLAGS+=-q mbcs -qlanglvl=extended -qinfo=pro
 else
   # xlclang options
   GLOBAL_CFLAGS+=-qlanglvl=extended0x -qxlcompatmacros
```
When inverting the code, the bodies if and else branches must be swapped as well.

keithc-ca commented on tools/configure:
Why is this file changing when there are no changes to `tools/configure.ac`?

midronij replied:
I think that may just be the result of generating the file with a newer version of `autoconf`. I noticed that the old one was built with v2.69, whereas I used v2.71. I didn't manually make any changes to `tools/configure` myself

keithc-ca replied:
Looking closer, I see the effects of changes to `omrmakefiles/confighelpers.m4`.

midronij commented:
@jdmpapin when you have a moment could you please start the CI tests?

jdmpapin commented:
Jenkins build all

midronij commented:
@jdmpapin looks like all the CI tests are passing! I think this is ready to merge now
